File: api/src/services/providers/theMovieDb.ts
import axios from 'axios';

import { IMoviesProvider } from "../../contracts/IMoviesProvider";
import { IMovie } from "../../models/IMovie";

import { IDiscoverMoviesResponse } from './theMovieDb.types';

export class TheMovieDb implements IMoviesProvider {
  async getMoviesForYear(year: number): Promise<IMovie[]> {
    const discoverResponse = await axios.get<IDiscoverMoviesResponse>('https://api.themoviedb.org/3/discover/movie/', {
      params: {
        api_key: process.env.THEMOVIEDB_API_KEY,
        certification_country: 'us',
        primary_release_year: year
      }
    });

    return discoverResponse.data.results.map(x => ({
      id: x.id,
      title: x.title,
      description: x.overview,
      image: `https://image.tmdb.org/t/p/w500${x.poster_path}`,
      date: new Date(x.release_date)
    }) as IMovie);
  }
}

File: api/lib/the_movie_db.rb
# frozen_string_literal: true

class TheMovieDB
  def initialize
    Tmdb::Api.key(ENV['THEMOVIEDB_API_KEY'])
    Tmdb::Api.language('en')
  end

  def top_movies_by_year(year)
    Tmdb::Discover.movie(year: year, region: 'us').results.map { |movie| format_movie(movie) }
  end

  private

  def format_movie(movie)
    {
      id: movie.id,
      title: movie.title,
      description: movie.overview,
      image: "https://image.tmdb.org/t/p/w500#{movie.poster_path}",
      date: movie.release_date
    }
  end
end
